### PR TITLE
Add tests on Lassen in Gitlab CI

### PR DESCRIPTION
**Pull request by adrienbernede**
This PR adds the capability to test MFEM on lassen in Gitlab CI.

Current stage:
- Adds basic cuda builds of MFEM (with no dependencies).
- Adds sanity checks.
- Tests on Lassen are triggered at the same time as tests on quartz and baseline tests (not sequentially).

TODO:
- [x] Fix minor failure in test builds (arch option power9 not recognized).
- [ ] Add some sort of baseline test on lassen if any.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1688](https://github.com/mfem/mfem/pull/1688) | @adrienbernede | @tzanio | @v-dobrev + @YohannDudouit + @camierjs + @jandrej | 08/13/20 | 09/01/20 | 09/08/20 | |

### CODE REVIEW

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
@tzanio @v-dobrev Here is the PR
@camierjs Can you help with the error in sanity-check? Seems to be pretty basic.
All I’m doing is:
```
make -j cuda
make -j test
```
As far as I remember, it used to work, but now I’m getting:
```
 cc1plus: error: unrecognized argument in option '-mcpu=power9'
 cc1plus: note: valid arguments to '-mcpu=' are: 401 403 405 405fp 440 440fp 464 464fp 476 476fp 505 601 602 603 603e 604 604e 620 630 740 7400 7450 750 801 821 823 8540 8548 860 970 G3 G4 G5 a2 cell e300c2 e300c3 e500mc e500mc64 e5500 e6500 ec603e native power3 power4 power5 power5+ power6 power6x power7 power8 powerpc powerpc64 powerpc64le rs64 titan
```

**Comment by v-dobrev** [comment]
The issue with the error about `-mcpu=power9` is due to old gcc. Loading a module with newer gcc should fix this.

Also, I noticed the build uses the `sm_60` arch (which is the default in the mfem build systems) -- this can be set at config time with the `make` option `CUDA_ARCH=sm_70` for Lassen.

**Comment by v-dobrev** [comment]
One other thing, I see that the Lassen tests run after the ones on Quartz. Can they be run in parallel?

**Comment by adrienbernede** [comment]
> One other thing, I see that the Lassen tests run after the ones on Quartz. Can they be run in parallel?

They effectively run in parallel. Just in another stage for clarity.

**Comment by adrienbernede** [comment]
LGTM

**Comment by dylan-copeland** [comment]
Note that PR #1584 introduces `-d cuda` unit tests when MFEM is built with CUDA, for TEST_CASEs with tag `[CUDA]`. Having tests run on lassen will be make those new unit tests more effective. More tests can be tagged with `[CUDA]` as appropriate.

File: .gitlab/lassen.yml [hunk 1]
```diff
@@ -20,38 +20,38 @@
...
 # Sanity check
 sanitycheck_mfem_ser_lassen:
-  extends: [.on_lassen]
+  extends: [.with_gcc_8_3_1, .on_lassen]
   stage: lassen_test
   needs: [build_mfem_ser_lassen]
   script:
     - cd ${BUILD_PATH}/${CI_PROJECT_NAME}_lassen_ser
-    - lalloc 1 -W 15 -q pdebug make -j test
+    - lalloc 1 -W 15 -q pdebug make -j test CUDA_ARCH=sm_70
```
Comment (v-dobrev) [inline .gitlab/lassen.yml:49]: No need to set `CUDA_ARCH` in the testing steps -- it is remembered from the configuration+build steps.

Comment (adrienbernede) [reply .gitlab/lassen.yml]: Done.

**Comment by adrienbernede** [comment]
LGTM

**Comment by v-dobrev** [comment]
Looks good now. There are two things that will be nice to add here:
* add an MPI build and test (`make test`),
* add some device runs (in `make test` and `ctest`) for examples and miniapps when cuda/hip is enabled.

**Comment by adrienbernede** [comment]
We could proceed with the PR as-is.
I won’t have the time to add lassen baseline-check before a while.

@v-dobrev adding MPI requires building dependencies, right?

**Comment by v-dobrev** [comment]
> We could proceed with the PR as-is.
> I won’t have the time to add lassen baseline-check before a while.

I agree. I won't be able to work on this right now either.

> @v-dobrev adding MPI requires building dependencies, right?

Yes, the parallel build requires hypre and metis.

**Comment by tzanio** [comment]
@YohannDudouit, @camierjs and @jandrej can you please test/review when you have a chance?

File: .gitlab/quartz.yml [hunk 2]
```diff
@@ -22,10 +22,6 @@
     MAKE_PAR: 6
     BASELINE_PAR: 18
 
-# Setup
-setup_quartz:
-  extends: [.setup, .on_quartz]
-
```
Comment (jandrej) [inline .gitlab/quartz.yml:28]: Why was this removed?

Comment (adrienbernede) [reply .gitlab/quartz.yml:28]: That’s a good question:
For now there is only a setup job for the part of the pipeline running on quartz (not for lassen).
I simplified it since it was unnecessarily convoluted. But it wasn’t a critical change.
In fact I may have to revert this change at some point when adding dependencies build on lassen...

**Comment by adrienbernede** [comment]
> I have not tested but this seems good as far as I understand, we could consider adding `MFEM_USE_CEED=YES` to the build options for testing the libCEED integration on GPU.

This would require building some dependencies first, right? That’s the next step :)

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
